serde/src/de/impls.rs

```diff
@@ -457,6 +655,45 @@ where
     }
 }
 
+impl<'a, 'de, T> Visitor<'de> for OptionFromVisitor<'a, T>
+where
+    T: Deserialize<'de>,
+{
+    type Value = ();
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("option")
+    }
+
+    #[inline]
+    fn visit_unit<E>(self) -> Result<(), E>
+    where
+        E: Error,
+    {
+        *self.0 = None;
+        Ok(())
+    }
+
+    #[inline]
+    fn visit_none<E>(self) -> Result<(), E>
+    where
+        E: Error,
+    {
+        *self.0 = None;
+        Ok(())
+    }
+
+    #[inline]
+    fn visit_some<D>(self, deserializer: D) -> Result<(), D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Some enum is opaque, can't do anything good here :(
```

Comment:
can't you do

```rust
match *self.0 {
    Some(ref mut data) => data.deserialize_from(deserializer),
    None => Some(T::deserialize(deserializer)?),
}
```
```

Reply:
Good point, but hard to say if this is justified. Very workload-specific.